### PR TITLE
Add logic to prevent changing user alias on logout/login

### LINKED ISSUE #921: Backend resets user alias and country on each login
If you assign yourself a custom alias via the dashboard, sign out (browser discards JWT), sign in again (requests new JWT), we re-request your public Steam info, and update your profile with it, *even if the user already exists*.

Doing this for avatars makes sense, but if the account already exists, we shouldn't be setting the alias or country code.

This is like a 2 line fix + tests and I already see the offending code, doing myself ASAP.

### PR DESCRIPTION
**Pull request by kanoka42**
Date:      Fri May 31 10:58:39 2024 -0500

Implements fix for bug #921

Implements fix for bug #921 which resets the users alias on logou and back in. This fix
    - Implements a new regex to disallow purely whitespace based aliases
    - E2e testing on the new aliases
    - Logic to test if the user name is valid, or is the default value

Closes #921

### CODE REVIEW

**Comment by kanoka42** [comment]
Closes #921

File: libs/constants/src/regexp/user-alias.regexp.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+export const USER_ALIAS_REGEXP = /^(?![\s028u{}]*$).+/;
```
Comment (GordiNoki) [inline libs/constants/src/regexp/user-alias.regexp.ts:1]: That's pretty weird regex. For example, nickname "u2u822" would be discarded, even though it is pretty random, it's not invalid

File: apps/backend/src/app/modules/users/users.service.ts [hunk 2]
```diff
@@ -162,10 +163,16 @@ export class UsersService {
       where: { steamID: userData.steamID }
     });
 
-    const input: Prisma.UserUpdateInput | Prisma.UserCreateInput = {
-      alias: userData.alias,
+    const isValidAlias = USER_ALIAS_REGEXP.test(user?.alias);
+
+    const findOrCreateUserData = {
```
Comment (GordiNoki) [inline apps/backend/src/app/modules/users/users.service.ts:168]: Why did you put that in a separate variable?

**Review by GordiNoki — CHANGES_REQUESTED** [review]
The commit name doesn't follow our naming conventions, make sure you have commit hooks working (run `npm install`). Also it worth doing test for the issue itself mentioned in #921, something like "should not update alias when user exists" in auth.e2e-spec.ts

File: apps/backend/src/app/modules/users/users.service.ts (hunk 2 above)
Comment (kanoka42) [reply apps/backend/src/app/modules/users/users.service.ts:168]: Just for read-ability, but I can move it inside the object if you want

**Review by tsa96 — APPROVED** [review]
Reviewed together in voice

**Review by tsa96 — APPROVED** [review]

**Comment by tsa96** [comment]
> The commit name doesn't follow our naming conventions, make sure you have commit hooks working (run `npm install`). Also it worth doing test for the issue itself mentioned in #921, something like "should not update alias when user exists" in auth.e2e-spec.ts

Some existing E2E tests were modified to check this, I think coverage is okay. Have asked for a unit test as well, then I think we're good.

**Comment by kanoka42** [comment]
> The commit name doesn't follow our naming conventions, make sure you have commit hooks working (run `npm install`). Also it worth doing test for the issue itself mentioned in #921, something like "should not update alias when user exists" in auth.e2e-spec.ts

The tests had to be reworked to account for this specific change, so a few of the tests check for this issue now.

**Comment by tsa96** [comment]
Woops!

File: libs/constants/src/regexp/user-alias.regexp.ts (hunk 1 above)
Comment (tsa96) [reply libs/constants/src/regexp/user-alias.regexp.ts]: Just to follow up on this - this was a regex I wrote, based on a SO answer (which I tried to dig up just now but failed). The intention is to match any string that's not *entirely* whitespace. I recommend regexr.com for if you're struggling to understand a regex, see [regexr.com/816p3](this one) for a breakdown of what this one is doing.

**Pull request merged**